src/main/kotlin/com/openlattice/mechanic/integrity/OrphanedLinkingPropertyData.kt

```diff
@@ -41,30 +47,71 @@ class OrphanedLinkingPropertyData(private val toolbox: Toolbox) : Check {
     }
 
     private fun tombstoneOrphanedLinkingProperties() {
-        executeUpdate(tombStoneSql, DeleteType.Soft)
+        logger.info("Starting task to clear orphaned linking property types values.")
+        val sw = Stopwatch.createStarted()
+
+        do {
+            sw.reset().start()
+
+            val count = toolbox.hds.connection.use { conn ->
+                conn.createStatement().use {
+                    it.executeUpdate(tombStoneSql)
+                }
+            }
+
+            logger.info(
+                    "Cleared $count orphaned linking property types in ${DATA.name} table in " +
+                            "${sw.elapsed(TimeUnit.MILLISECONDS)} ms."
+            )
+        } while (count > 0)
     }
 
     private fun deleteOrphanedLinkingProperties() {
-        executeUpdate(deleteSql, DeleteType.Hard)
-    }
+        logger.info("Starting task to delete orphaned linking property types values.")
+        val sw = Stopwatch.createStarted()
 
-    private fun executeUpdate(sql: String, deleteType: DeleteType) {
-        val deleteTypeStr1 = if (deleteType == DeleteType.Hard) "delete" else "clear"
-        val deleteTypeStr2 = if (deleteType == DeleteType.Hard) "Deleted" else "Cleared"
+        do {
+            sw.reset().start()
+            var count = 0L
 
-        val sw = Stopwatch.createStarted()
-        logger.info("Starting task to $deleteTypeStr1 orphaned linking property types values.")
-        val count = toolbox.hds.connection.use { conn ->
-            conn.createStatement().use {
-                it.executeUpdate(sql)
+            BasePostgresIterable(
+                    StatementHolderSupplier(toolbox.hds, selectDeletableSql)
+            ) { rs ->
+                LinkedEntityRow(
+                        ResultSetAdapters.id(rs), ResultSetAdapters.originId(rs), ResultSetAdapters.propertyTypeId(rs)
+                )
             }
-        }
-        logger.info(
-                "$deleteTypeStr2 $count orphaned linking property types in ${DATA.name} table in " +
-                        "${sw.elapsed(TimeUnit.MILLISECONDS)} ms.")
+                    .groupBy { it.linkingId }
+                    .forEach { (linkingId, deletablesOfLinkingId) ->
+                        deletablesOfLinkingId
+                                .map { it.originId to it.propertyTypeId }
+                                .groupBy { it.first } // originId
```

Comment:
FWIW you don't need to make map out of this, just .groupBy(it.originId) and then propertyTypeIdsByOriginIds below will be a collection of LinkedEntityRows, so line 93 would be .map { it.propertyTypeId } instead.

Reply:
Ah yeah, true! Thanks!